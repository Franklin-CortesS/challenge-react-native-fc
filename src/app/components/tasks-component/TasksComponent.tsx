
import { Button } from "@react-native-elements/base";
import React, { useState } from "react";
import { View, StyleSheet, Modal, Text, TextInput } from 'react-native';
import TaskListComponent from "../task-list-component/TaskListComponent";
import { connect } from 'react-redux';
import { addTasks } from "../../store/reducers/TaskReducer";

const mapStateToProps = (state: any) => {
    return {
      tasks: state
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return {
      addTask: (obj: any) => dispatch(addTasks(obj)),
    };
};

const TasksComponent = (props: any) => {
    const [text, onChangeText] = useState("");
    const [modalVisible, setModalVisible] = useState(false);
    const [addError, onChangeDisplayError] = useState(false);

    const saveTask = () => {
        onChangeDisplayError(false);

        if(text.length > 0){
            props.addTask({
                id: Math.floor(Math.random() * 1000),
                task: text,
            });
            setModalVisible(!modalVisible);
            onChangeText('');
        }
        else{
            onChangeDisplayError(true);
        }
    }

    const closeAdd = () => {
        onChangeDisplayError(false);
        setModalVisible(!modalVisible);
        onChangeText('');
    }
    
    return (
        <View style={styles.container}>
            <Modal
                animationType="slide"
                transparent={true}
                visible={modalVisible}
                onRequestClose={() => {
                    setModalVisible(!modalVisible);
                }}
            >
                <View style={styles.centeredView}>
                    <View style={styles.modalView}>
                        <TextInput
                            style={styles.input}
                            onChangeText={onChangeText}
                            value={text}
                            placeholder="Tarea"
                            placeholderTextColor="#000000"
                        />
                        {addError == true && <Text style={styles.errorText}>Escribe la tarea a guardar.</Text>}
                        <View style={styles.twoBtns}>
                            <Button
                                buttonStyle={[styles.button, styles.buttonClose]}
                                onPress={() => closeAdd()}
                                title="Cerrar"
                            >
                            </Button>
                            <Button
                                buttonStyle={[styles.button, styles.buttonSave]}
                                onPress={() => saveTask()}
                                title="Agregar"
                            >
                            </Button>
                        </View>
                    </View>
                </View>
            </Modal>

            <Button
            buttonStyle={styles.my_button}
            onPress={() => setModalVisible(!modalVisible)}
            title="Nueva tarea"></Button>

            <TaskListComponent tasks={props.tasks}></TaskListComponent>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        color: "#000000"
    },
    my_button: {
        marginTop: 15,
        marginBottom: 5,
        marginHorizontal: 20,
        backgroundColor: "#009900",
    },
    centeredView: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        marginTop: 22
      },
      modalView: {
        margin: 20,
        backgroundColor: "white",
        borderRadius: 20,
        padding: 35,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: {
          width: 0,
          height: 2
        },
        shadowOpacity: 0.25,
        shadowRadius: 4,
        elevation: 5
      },
      button: {
        borderRadius: 20,
        padding: 10,
      },
      buttonSave: {
        backgroundColor: "#009900",
        marginLeft: 2
      },
      buttonClose: {
        backgroundColor: "#CC0000",
        marginRight: 2
      },
      modalText: {
        marginBottom: 15,
        textAlign: "center",
        color: "#000000"
      },
      input: {
        height: 40,
        marginHorizontal: 12,
        borderWidth: 1,
        padding: 10,
        color: "#000000",
      },
      twoBtns: {
          display: "flex",
          flexDirection: 'row',
          justifyContent: 'space-between',
          marginTop: 12
      },
      errorText: {
        color: "#CC0000",
        fontSize: 10,
      },
});

export default connect(mapStateToProps, mapDispatchToProps) (TasksComponent);