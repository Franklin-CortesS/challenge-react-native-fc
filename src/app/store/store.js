import { configureStore } from "@reduxjs/toolkit";
import { reducer } from "./reducers/TaskReducer";

const store = configureStore({
    reducer: reducer
})

export default store;