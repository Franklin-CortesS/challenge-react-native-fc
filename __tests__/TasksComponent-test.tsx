/**
 * @format
 */

 import 'react-native';
 import React from 'react';
 
 // Note: test renderer must be required after react-native.
 import renderer from 'react-test-renderer';
 import TasksComponent from '../src/app/components/tasks-component/TasksComponent';
 import { Provider } from 'react-redux';
 import store from '../src/app/store/store';
 import { render, fireEvent, waitFor } from "@testing-library/react-native";
 
  jest.useFakeTimers();

  describe("Tests for TaskComponent", () => {
    it('renders correctly', () => {
      const tree = renderer.create(<Provider store={store}>
        <TasksComponent />
      </Provider>).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('open and close modal', () => {
      const { getByText} = render(<Provider store={store}>
        <TasksComponent />
      </Provider>);

      //open modal
      const modal = getByText('Nueva tarea');
      fireEvent.press(modal);

      //close modal
      const closemodal = getByText('Cerrar');
      fireEvent.press(closemodal);
    })

    test('add new task', () => {
      const { getByText, getByPlaceholderText } = render(<Provider store={store}>
        <TasksComponent />
      </Provider>);

      //open modal
      const modal = getByText('Nueva tarea');
      fireEvent.press(modal);

      //test add
      const input = getByPlaceholderText("Tarea");
      expect(input).toBeTruthy();
      const task = "Tarea 1";
      fireEvent.changeText(input, task);
      fireEvent.press(getByText('Agregar'));

      const tasks: any = store.getState();
      expect(tasks.length).toEqual(1);
      expect(tasks).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            task: task,
          }),
        ])
      );
    })

    test('add new error task', () => {
      const { getByText, getByPlaceholderText } = render(<Provider store={store}>
        <TasksComponent />
      </Provider>);

      //open modal
      const modal = getByText('Nueva tarea');
      fireEvent.press(modal);

      //test add
      const input = getByPlaceholderText("Tarea");
      expect(input).toBeTruthy();
      const task = "";
      fireEvent.changeText(input, task);
      fireEvent.press(getByText('Agregar'));

      expect(getByText('Escribe la tarea a guardar.')).toBeTruthy();
    })
  })

 