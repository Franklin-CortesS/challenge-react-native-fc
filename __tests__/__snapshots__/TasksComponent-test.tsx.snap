// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TaskComponent renders correctly 1`] = `
<View
  style={
    Object {
      "color": "#000000",
      "flex": 1,
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#009900",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 5,
            "marginHorizontal": 20,
            "marginTop": 15,
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Nueva tarea
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#E5E5E5",
              }
            }
          >
            No hay tareas que mostrar.
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
